libraries/Microsoft.Bot.Builder.Core/Middleware/StateMiddleware.cs

```diff
@@ -0,0 +1,174 @@
+    }
+
+    public abstract class StateMiddleware<StateT> : IContextCreated, ISendActivity
+        where StateT : IStoreItem, new()
+    {
+        private readonly StateSettings _settings;
+        private readonly IStorage _storage;
+
+        /// <summary>
+        /// Create statemiddleware
+        /// </summary>
+        /// <param name="name">name of the kind of state</param>
+        /// <param name="storage">storage provider to use</param>
+        /// <param name="settings">settings</param>
+        public StateMiddleware(IStorage storage, StateSettings settings = null)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+            _settings = settings ?? new StateSettings();
+        }
+
+        public abstract string GetPropertyName();
+
+        /// <summary>
+        /// The key extracted from the activity which is used to store/retrieve the state
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <returns></returns>
+        public abstract string GetStorageKey(IBotContext context);
+
+        public async Task ContextCreated(IBotContext context, MiddlewareSet.NextDelegate next)
+        {
+            await Read(context).ConfigureAwait(false);
+            await next().ConfigureAwait(false);
+        }
+
+        public async Task SendActivity(IBotContext context, IList<Activity> activities, MiddlewareSet.NextDelegate next)
+        {
+            if (_settings.WriteBeforeSend)
+            {
+                await Write(context).ConfigureAwait(false);
+            }
+            await next().ConfigureAwait(false);
+            if (!_settings.WriteBeforeSend)
+            {
+                await Write(context).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async Task<StoreItems> Read(IBotContext context)
+        {
+            var key = this.GetStorageKey(context);
+            var keys = new List<String>();
+            keys.Add(key);
+            var items = await _storage.Read(keys.ToArray());
+            var state = items.Get<StateT>(key);
+            if (state == null)
+                state = new StateT();
+            context.Set(GetPropertyName(), state);
+            return items;
+        }
+
+        protected virtual async Task Write(IBotContext context, StoreItems changes = null)
+        {
+            if (changes == null)
+                changes = new StoreItems();
+
+            var state = context.Get<StateT>(GetPropertyName());
+            if (state == null)
+                state = new StateT();
+            var key = GetStorageKey(context);
+            changes[key] = state;
+
+            if (this._settings.LastWriterWins)
+            {
+                foreach (var item in changes)
+                {
+                    ((StoreItem)changes[item.Key]).eTag = "*";
+                }
+            }
+
+            await _storage.Write(changes).ConfigureAwait(false);
```

Comment:
>_storage.Write(changes) [](start = 18, length = 23)

I maintain a change hash and only commit the changes should something actually change. If you're going to support writing before send (I don't currently) you definitely need that otherwise you'd write multiple times per turn. Its a good optimization to have anyway.